src/base/Connection.cpp

```diff
@@ -16,80 +16,46 @@ Connection::~Connection() {
 }
 
 inline bool isSkippableError(int err_no) {
-  return (err_no == ECONNRESET || err_no == ETIMEDOUT ||
+  return (err_no == EAGAIN || err_no == ECONNRESET || err_no == ETIMEDOUT ||
           err_no == EWOULDBLOCK || err_no == EHOSTUNREACH || err_no == EPIPE ||
           err_no == EBADF  // Bad file descriptor can happen when
-                           // there's a race condition between ta thread
+                           // there's a race condition between a thread
                            // closing a connection and one
                            // reading/writing.
   );
 }
 
-ssize_t Connection::read(string* buf) {
+bool Connection::read(string* buf) {
   VLOG(4) << "Before read get connectionMutex";
   lock_guard<std::recursive_mutex> guard(connectionMutex);
   VLOG(4) << "After read get connectionMutex";
-  // 2s should be enough since EAGAIN is rare in blocking socket.
-  int CLIENT_timeout = 2;
-  // Try at 10Hz
-  int num_trails = CLIENT_timeout * 10;
-  for (int trials = 0; trials < num_trails; trials++) {
-    ssize_t messagesRead = reader->read(buf);
-    if (messagesRead == -1) {
-      if (trials < (num_trails - 1) && errno == EAGAIN) {
-        // If we get EAGAIN, assume the kernel needs to finish
-        // flushing some buffer and retry after a delay.
-        usleep(100000);
-        LOG(INFO) << "Got EAGAIN, waiting 100ms...";
-      } else if (trials == (num_trails - 1) && errno == EAGAIN) {
```

Comment:
This was added to address the EAGAIN case which sometimes show up in the reconnect. I'm not sure if our refactor resolves this issue. 

I will run this branch in my environment to test after this PR is merged. :)

Reply:
I think I figured out what was happening: BackedReader did not have the same locks that BackedWriter had, so it's possible to have a race condition.  The global "errno" variable just has the most recent error so if we ever run two unix commands at the same time, we risk clobbering it.  That's why we have to be so strict with the mutexes :-)